Swift Archival & Serialization API

itaiferber opened the pull request:
**What's in this pull request?**
Now that the core types introduced by [SE-0166](https://github.com/apple/swift-evolution/blob/master/proposals/0166-swift-archival-serialization.md) have been sunk into the Swift standard library, the stdlib and compiler portions of #8124 and #8125 can be combined into one PR (this one).

The features can be tested together; #8125 was previously not tested on its own since it required the presence of #8124.

The Foundation portions of #8124 will be put up in a subsequent PR.

rdar://problem/30472044

Commit:
Integrate *Codable protocols and associated types

Add Encodable and Decodable protocols, along with associated container
types, default implementations, and extensions to the standard library

Commit:
Add CodingKey and *codable derived conformance

* Allow CodingKey conformance to be automatically derived for enums
  which have no raw type (with no associated values) and which have
  a raw type of String or Int.
* Allow Encodable and Decodable conformance to be automatically derived
  for classes and structs with Encodable/Decodable properties
* Add initial unit tests for verifying derived conformance

Commit:
CodingKey/*Codable tests don't need ObjC interop

Since all of the core types are now in the Swift stdlib, the tests no
longer rely on objc_interop or Foundation.

itaiferber commented:
@swift-ci Please test

itaiferber commented:
I'm not sure these test failures are actually related to my changes. On Linux we're getting failures like `error: 'Iterator' is not a member type of 'Sequence'`, and on macOS, the failing tests are SourceKit. The Linux failures are more worrisome to me though; anyone know what's going on there?

Commit:
Remove unnecessary enum case check

itaiferber commented:
Let's try again

itaiferber commented:
@swift-ci Please test

obitow commented on stdlib/public/core/Codable.swift:
```diff
@@ -0,0 +1,3140 @@
+    /// - throws: `DecodingError.typeMismatch` if the encountered encoded value is not convertible to the requested type.
+    /// - throws: `DecodingError.keyNotFound` if `self` does not have an entry for the given key.
+    /// - throws: `DecodingError.valueNotFound` if `self` has a null entry for the given key.
+    func decode<T : Decodable>(_ type: T.Type, forKey key: Key) throws -> T
```
@itaiferber great proposal, can't wait to start using this. But I have a question: wouldn't be better to drop the type from the parameters and just use inference here? Something like:
```
func decode<T : Decodable>(_ key: Key) throws -> T
```

itaiferber replied:
@obitow Thanks!

This point has been discussed at length on swift-evolution; dropping the type parameter can lead to ambiguous statements, which can be confusing and unintuitive. Please feel free to refer to those discussions for the full story.

itaiferber commented:
@atrick @gottesmm Looks like the tests are failing in SIL validation on Linux and the errors are kind of bizarre: e.g. `error: 'Iterator' is not a member type of 'Sequence'`. Any idea about what's going on here/how I can address this?

itaiferber commented:
Let's try the smoke tests at least?

itaiferber commented:
Perhaps we can try the smoke tests at least?

itaiferber commented:
@swift-ci Please smoke test

Commit:
Rename sequence type to avoid ambiguity

itaiferber commented:
It's possible that `Sequence : Swift.Sequence` was causing ambiguity when generating SIL. I've changed that; let's see if that's caused any changes in the tests. Once more with vigor...

itaiferber commented:
@swift-ci Please test

itaiferber commented:
Looks like the `Sequence` change did fix the SIL tests. The SourceKit tests are still failing though.

Commit:
Derive encoding an empty container by default

struct X : Codable {} should generate an encode(to:) which encodes an
empty keyed container, not an encode(to:) which does nothing

Commit:
Fix crash in AST printing

Extensions need not have nominal types, e.g.

    extension Int : Codable { ... }

where

    typealias Codable = Encodable & Decodable

Commit:
Fix init(from:) required condition

itaiferber commented:
@swift-ci Please test

itaiferber commented:
`bdfa927` should have hopefully fixed the SourceKit and `swift-ide-test` issues as `ASTPrinter` was making assumptions about extensions that `typealias Codable = Encodable & Decodable` broke

itaiferber commented:
Looks like the macOS build failure is not unique to this PR. (See also https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RDA_test-simulator/3143/consoleFull)
Waiting for a resolution on the issue and I'll re-test on macOS.

itaiferber commented:
Waiting on #9081 to go in before we can retest

itaiferber commented:
Should be good to go, hopefully

itaiferber commented:
@swift-ci Please test

rintaro commented on stdlib/public/core/Codable.swift:
```diff
@@ -0,0 +1,3140 @@
+    ///
+    /// - parameter encoder: The encoder to write data to.
+    /// - throws: An error if any values are invalid for `encoder`'s format.
+    func encode(to encoder: Encoder) throws
```
Out of curiosity:
Why this is not `func encode<E : Encoder>(to encoder: E) throws`?
I believe generics have more optimization opportunities than protocols existentials.

itaiferber replied:
@rintaro `encode(to:)` and `init(from:)` are not generic because there are cases where you need to pass an existential `Encoder` or `Decoder` to an object. For instance, given:

```swift
class A : Codable { /* ... */ }
class B : A { /* ... */ }
```

`class B` would likely implement `encode(to:)` as

```swift
func encode(to encoder: Encoder) throws {
    var container = encoder./* some container type */
    // encode properties into container

    super.encode(to: container.superEncoder())
}
```

This is not possible to do if `encode(to:)` is generic, because `superEncoder()` cannot return a concrete type, and `encode<E : Encoder>(to encoder: E)` cannot take an existential type.

rintaro replied:
Ah, I've forgotten that `<T : P>` can't receive `P`.
Thank you for clarifying!

itaiferber commented:
I'm wondering if this macOS failure is extant, or actually related to any of my stuff.

itaiferber commented:
@swift-ci Please test macOS

DougGregor commented:
It looks like we're getting some duplicate declarations somehow in that SourceKit failure:

    23:37:31              key.kind: source.lang.swift.ref.function.constructor,
    23:37:31              key.name: "init(rawValue:)",
    23:37:31              key.usr: "s:s16RawRepresentablePxSg0A5ValueQz03rawC0_tcfc"
    23:37:31 +          },
    23:37:31 +          {
    23:37:31 +            key.kind: source.lang.swift.ref.function.constructor,
    23:37:31 +            key.name: "init(rawValue:)",
    23:37:31 +            key.usr: "s:s16RawRepresentablePxSg0A5ValueQz03rawC0_tcfc"
    23:37:31            }

itaiferber commented:
@DougGregor Yeah, I'm investigating why this happens — seems weird considering my stuff doesn't touch these types at all, or defines anything with `init(rawValue:)`. This reproduces locally with my changes, so it's definitely a present issue; trying to see if I can reproduce it without my changes, and in parallel, see if I can track down what about SourceKit is causing this to happen. `sourcekitd-test` connects to SourceKit over XPC, so I want to find out what instance of the daemon it's connecting to; a mismatch may cause weird behavior...

itaiferber commented:
Not sure how SourceKit could see two `init(rawValue:)` constructors to begin with, since that shouldn't be possible to declare.

itaiferber commented:
It looks like this change is caused by the inclusion of `extension RawRepresentable where RawValue == ..., Self : Codable { ... }`; commenting out those extensions prevents the repeated conformance.

That being said, I don't know, then, if this SourceKit output is expected (in which case I will update the test to match), or if this is a SourceKit bug, or otherwise. @DougGregor Any ideas? Who can I consult with about this?

DougGregor commented:
Weird, because we're seeing two entities with exactly the same USR (`s16RawRepresentablePxSg0A5ValueQz03rawC0_tcfc`), which is really odd. Perhaps a consequence of projecting the protocol extensions onto the concrete type. @akyrtzi or @nkcsgexi , any ideas?

DougGregor commented:
It seems like a bug like this in the actual AST (e.g., duplicated initializers) would cause everything to break horribly, so it feels like a SourceKit bug.

Commit:
Split RawRep extensions for Encodable + Decodable

You shouldn't need to be Codable in order to get these implementations;
being either Encodable or Decodable should be enough to get the relevant
default implementation

itaiferber commented:
This last commit doesn't fix the issue, but if I'm already looking at the extensions... (Let's split them up so you can benefit even by just being `Encodable` or `Decodable`.)

itaiferber commented on stdlib/public/core/Codable.swift:
```diff
@@ -1833,6 +1833,14 @@ public struct CodingUserInfoKey : RawRepresentable, Equatable, Hashable {
+    ///
+    /// - parameter lhs: The key to compare against.
+    /// - parameter rhs: The key to compare with.
+    public static func ==(_ lhs: CodingUserInfoKey, _ rhs: CodingUserInfoKey) -> Bool {
```
I'd meant to split this out into a different commit, but it made it in here anyway. `CodingUserInfoKey` was missing `==`, but the compiler didn't complain.

itaiferber commented:
Per @akyrtzi's suggestion via email, going to update the test to prevent this from blocking us and file a Radar about it.

Commit:
Temporarily fix failing SourceKit test

itaiferber commented:
@swift-ci Please test and merge

itaiferber commented:
@swift-ci Please test

itaiferber commented:
Build failed with an unrelated issue that's been reverted; let's try again.

itaiferber commented:
@swift-ci Please test

itaiferber commented:
For interested Apple folks, BTW, rdar://31896858

itaiferber commented:
Looks like a Migrator test is failing on iOS: `Migrator/wrap_optional.swift`
Looking into it.

itaiferber commented:
The builder is producing a module file for the incorrect target? `error: module file was created for incompatible target x86_64-apple-macosx10.9`

itaiferber commented:
@shahmishal @DougGregor This... Seems wrong. Any idea why this might happen?

itaiferber commented:
Seems like https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx/ failed with this as well...

itaiferber commented:
Waiting on #9113 before testing again.

itaiferber commented:
Let's try that again

itaiferber commented:
@swift-ci Please test

itaiferber commented:
Excellent. @parkera @DougGregor If either of sees this, can you please merge it in?

itaiferber commented:
Thanks @DougGregor!

norio-nomura commented on stdlib/public/core/Codable.swift:
```diff
@@ -0,0 +1,3176 @@
+
+// An implementation of _KeyedEncodingContainerBase and _KeyedEncodingContainerBox are given at the bottom of this file.
+/// `KeyedEncodingContainer` is a type-erased box for `KeyedEncodingContainerProtocol` types, similar to `AnyCollection` and `AnyHashable`. This is the type which consumers of the API interact with directly.
+public struct KeyedEncodingContainer<K : CodingKey> {
```
@itaiferber Should `KeyedEncodingContainer` declare` KeyedEncodingContainerProtocol` conformance?

itaiferber replied:
It can, but I don't see much of a benefit to doing so. When would you want to use it as `KeyedEncodingContainerProtocol` instead of the fully-typed `KeyedEncodingContainer<K>`?

norio-nomura replied:
Indeed, you are right. I can not think of such a scene.
I just wondered watching `KeyedDecodingContainer` declared as`KeyedDecodingContainerProtocol`.

itaiferber replied:
Good catch — I don't think that's really necessary either. It doesn't hurt, but we should be consistent. I'll consider this; thanks!

norio-nomura commented on stdlib/public/core/Codable.swift:
```diff
@@ -0,0 +1,3176 @@
+
+    /// Decodes a value of the given type for the given key, if present.
+    ///
+    /// This method retu
```
@itaiferber Why is it `mutating` only for `decode` of `UnkeyedDecodingContainer`?

itaiferber replied:
`UnkeyedDecodingContainer`s need to maintain some form of state to keep track of where they are in decoding, whether an index into an array that's incrementing with every `decode(...)` call, an offset into a data blob or stream, etc. The methods are `mutating` to indicate that, and to allow the implementation to be a `struct`. This is as opposed to `KeyedDecodingContainer`s and `SingleValueDecodingContainer`s, which by design will almost certainly not need to do that. Is there a use case you have in mind for making those `mutating` as well?

norio-nomura replied:
Certainly, I could not think of any reason to make `decode()` of another protocol `mutating` as well.
Thank you for answering my boring questions. :pray:

itaiferber replied:
Not at all! Thank you for paying such close attention to this; it's the best way for us to get direct feedback on it, and I appreciate your efforts.

norio-nomura commented:
I filed [SR-4861](https://bugs.swift.org/browse/SR-4861) that request adding `encode<T>(_:)` and `decode<T>(_:)` to `SingleValueEncodingContainer` and `SingleValueDecodingContainer`.

itaiferber commented:
@norio-nomura Thanks, assigned that to myself. That change, among a few others, is planned, and going through some API review soon.

Jceciliani commented on test/decl/protocol/special/coding/class_codable_inheritance.swift:
```diff
@@ -0,0 +1,55 @@
+    let _ = SimpleChildClass.CodingKeys.self // expected-error {{'CodingKeys' is inaccessible due to 'private' protection level}}
+
+    // The enum should have a case for each of the vars.
+    // NOTE: This expectedxerror will need to be removed in the future.
```
Small text fix needed in comment. "//NOTE: This expected error will need to be removed in the future."

itaiferber replied:
Thanks! #16749

Jceciliani reviewed (commented):
All committed changes appear correct after review.
